clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -559,13 +559,14 @@ private List<ProducerBatch> drainBatchesForOneNode(Cluster cluster, Node node, i
         int size = 0;
         List<PartitionInfo> parts = cluster.partitionsForNode(node.id());
         List<ProducerBatch> ready = new ArrayList<>();
-        /* to make starvation less likely this loop doesn't start at 0 */
+        /* to make starvation less likely each node has it's own drainIndex */
+        int drainIndex = getDrainIndex(node.idString());
         int start = drainIndex = drainIndex % parts.size();
         do {
             PartitionInfo part = parts.get(drainIndex);
             TopicPartition tp = new TopicPartition(part.topic(), part.partition());
-            this.drainIndex = (this.drainIndex + 1) % parts.size();
-
+            updateDrainIndex(node.idString(), drainIndex);
```

Comment:
I still think we can updateDrainIndex after while loop. After all, we only use the `drainIndex` variable inside while loop, not related to `nodesDrainIndex` map, right?

Reply:
the drainIndex should be related to tp, take a look at the following example(node has to tp, tp-0  tp-1)

// drainIndex  = 1
drainIndex = getDrainIndex();
do {
   drainIndex = 1 ;
   tp = tp-1;
   // update drainIndex to 0 without updateDrainIndex
   drainIndex = 0 ;
   synchronized (deque) {
       // satisfy the break condition
      break;
    }
}
while (start != drainIndex)
// record  drainIndex = 0;    (the drainIndex should be 1 right ? so the next drainBatch will begin with tp-1)
updateDrainIndex()

Reply:
Thanks for the explanation. Make sense.